pages/common/views/components/datatable.jsx

```diff
@@ -1,48 +1,125 @@
+class Table extends Component {
+  constructor(options) {
+    super(options);
+    this.toggleContent = this.toggleContent.bind(this);
+    this.isExpandable = this.isExpandable.bind(this);
   }
-  const columns = !isEmpty(schema)
-    ? schema
-    : Object.keys(data[0]).reduce((obj, key) => ({ ...obj, [key]: {} }), {});
-  return (
-    <table className="govuk-react-datatable">
-      <thead>
-        <tr>
+
+  componentDidMount() {
+    if (!this.props.expandable) {
+      return;
+    }
+    const expanded = this.props.data.reduce((obj, { id }) => {
+      return { ...obj, [id]: false };
+    }, {});
+    this.setState({ expanded });
+  }
+
+  toggleContent(id) {
+    const { expanded } = this.state;
+    expanded[id] = !expanded[id];
+    this.setState({ expanded });
+  }
+
+  isExpandable(id) {
+    if (this.props.expandable) {
+      if (!this.state) {
+        return true;
+      }
+      return this.state && this.state.expanded[id];
+    }
+  }
+
+  render() {
+    const {
+      data,
+      schema,
+      sortable,
+      expandable,
+      url,
+      editLink = 'Change',
+      deleteLink = 'Remove'
+    } = this.props;
+    if (isUndefined(data)) {
+      throw new Error('data must be provided');
+    }
+    const columns = !isEmpty(schema)
+      ? pickBy(schema, s => !s.expandable)
+      : Object.keys(data[0]).reduce((obj, key) => ({ ...obj, [key]: {} }), {});
+    const expandableData = pickBy(schema, s => s.expandable);
+    return (
+      <table className="govuk-react-datatable">
+        <thead>
+          <tr>
+            {
+              map(columns, (column, key) =>
+                <DatatableHeader key={key} id={key} sortable={sortable} { ...column } />
+              )
+            }
+          </tr>
+          <tr id="filter-header"></tr>
+        </thead>
+        <tbody>
           {
-            map(columns, (column, key) =>
-              <DatatableHeader key={key} id={key} sortable={sortable} { ...column } />
-            )
+            data.map(row => (
+              <Fragment key={row.id}>
+                <tr
+                  onClick={ expandable && (() => this.toggleContent(row.id)) }
+                  className={classnames({
+                    expandable,
+                    restrictions: row.notes,
+                    expanded: expandable && this.state && this.state.expanded[row.id]
+                  })}
+                >
+                  {
+                    map(columns, (schema, key) => {
+                      const datum = getValue({ row, schema, key });
+                      return <td key={key} className={key}>
+                        { schema.format ? schema.format(datum, row) : datum }
+                      </td>;
+                    })
+                  }
+                </tr>
+                {
+                  this.isExpandable(row.id) && (
+                    <tr className='expanded-content' onClick={ expandable && (() => this.toggleContent(row.id)) }>
```

Comment:
Is it correct that clicking the _expanded_ row should un-expand it? Doesn't seem right to me.

Reply:
which means, if there is no visual separation between the new and standard content, the user doesn't have any visual cue as to where they click to collapse the row again

Reply:
Fair. Leave it as it is then. It's not like it's hard to change in future if we want to.